src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,190 @@
+        }
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        Node<T> target = getNode(index);
+        linkBefore(value, target);
+    }
+
+    private void linkBefore(T value, Node<T> newNodeBefore) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (newNodeBefore == head) {
+            newNode.next = newNodeBefore;
+            newNodeBefore.prev = newNode;
+            head = newNode;
+            return;
+        }
+        Node<T> previousNode = newNodeBefore.prev;
+        newNode.next = newNodeBefore;
+        newNodeBefore.prev = newNode;
+        previousNode.next = newNode;
+        newNode.prev = previousNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        exceptionForIndex(index);
+        Node<T> currentNode = getNode(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        exceptionForIndex(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        exceptionForIndex(index);
+        Node<T> currentNode;
+        currentNode = getNode(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size(); i++) {
+            if (Objects.equals(object, currentNode.value)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> node) {
+        if (size() == 1) {
+            head = new Node<>(null, null, null);
+            return;
+        }
+        if (node == head) {
+            head = head.next;
+            node.next.prev = null;
+            return;
+        }
+        if (node == tail) {
+            node.prev.next = null;
+            return;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            int middleSize = size() / 2;
+            if (middleSize < index) {
+                currentNode = tail;
+                for (int j = size() - 1; j > index; j--) {
+                    currentNode = currentNode.prev;
+                }
+            } else {
+                currentNode = head;
+                for (int j = 0; j < index; j++) {
+                    currentNode = currentNode.next;
+                }
+            }
+        }
```

Comment:
```suggestion
        Node<T> currentNode = head;
        int middleSize = size() / 2;
        if (middleSize < index) {
            currentNode = tail;
            for (int j = size() - 1; j > index; j--) {
                currentNode = currentNode.prev;
            }
        } else {
            currentNode = head;
            for (int j = 0; j < index; j++) {
                currentNode = currentNode.next;
            }
        }
```
you dont need to iterate index times, to find the node by index